### PR TITLE
Add `IntoIterator` for `Box<[T]>` + edition 2024-specific lints

### PR DESCRIPTION
**Pull request by compiler-errors**
* Adds a similar method probe opt-out mechanism to the `[T;N]: IntoIterator` implementation for edition 2021.
* Adjusts the relevant lints (shadowed `.into_iter()` calls, new source of method ambiguity).
* Adds some tests.
* Took the liberty to rework the logic in the `ARRAY_INTO_ITER` lint, since it was kind of confusing.

Based mostly off of #116607.

ACP: rust-lang/libs-team#263
References #59878
Tracking for Rust 2024: https://github.com/rust-lang/rust/issues/123759

Crater run was done here: https://github.com/rust-lang/rust/pull/116607#issuecomment-1770293013
Consensus afaict was that there is too much breakage, so let's do this in an edition-dependent way much like `[T; N]: IntoIterator`.

### CODE REVIEW

File: compiler/rustc_ast/src/ptr.rs
Comment (compiler-errors) [inline compiler/rustc_ast/src/ptr.rs:187]: Very funny to see somewhere in the compiler where we were relying on this behavior.

**Comment by tmandry** [comment]
I looked over the PR and I think we can kick off FCP in tandem with code review. The implementation includes a full migration that behaves as I'd expect.

@rust-lang/lang Are we okay with applying the same method resolution hack used for slices in 2021 to boxed slices in 2024?

@rust-lang/libs-api Are we okay with insta-stabilizing `impl<T> IntoIterator for Box<[T]>` and applying the method resolution hack?

@rfcbot fcp merge

**Comment by BurntSushi** [comment]
The `IntoIterator` impl using `vec::IntoIter<I, A>` for `Box<[T]>` strikes me as a little odd, since `Box<[T]>` isn't a `Vec`. But I also can't necessarily come up with a specific concern with reusing the type other than "it seems weird."

**Comment by compiler-errors** [comment]
For FCP folks, see some of the discussion in #116607 -- apparently reusing `vec::IntoIter` was discussed in a T-libs-api meeting?

---

Also, this PR doesn't implement `IntoIterator for Box<[T; N]>`. That impl seems slightly less useful, since calling `Box::new([1, 2, 3]).into_iter()` already uses the `IntoIterator for [T; N]` impl via autoderef, so we already get an `impl Iterator<Item = T>`.

If anyone has a strong opinion, I could either add to this PR, or (ideally) add to this PR in a follow-up. It would only cause a breaking change when the user is relying on the exact `array::Iter` output type of `[T; N]`, rather than the much larger fallout of this change.

**Comment by clarfonthey** [comment]
> Also, this PR doesn't implement IntoIterator for Box<[T; N]>. That impl seems slightly less useful, since calling Box::new([1, 2, 3]).into_iter() already uses the IntoIterator for [T; N] impl via autoderef, so we already get an impl Iterator<Item = T>.
>
> If anyone has a strong opinion, I could either add to this PR, or (ideally) add to this PR in a follow-up. It would only cause a breaking change when the user is relying on the exact array::Iter output type of [T; N], rather than the much larger fallout of this change.

My personal opinion here is that it would be desirable to have a heap-based iterator here, since using the array implementation would copy the entire array onto the stack *before* iterating. It also helps avoid the weird discrepancy between `Box<[T]>` and `Box<[T; N]>` having different performance implications.

**Comment by compiler-errors** [comment]
I've kicked off an experiment to see if `Box<[T; N]>: IntoIterator` needs a hack like this, or if we can get away with just adding the impl: #124108.

Let's not let that experiment block this change + FCP; I can generalize the hacks being added here in a subsequent PR if we see that it does need a method dispatch hack too, and that can be FCP'd separately; since this isn't really concerned with any open questions, I don't expect it to be particularly risky to separate into two pieces for edition 2024.

File: src/tools/clippy/tests/ui/into_iter_on_ref.fixed
Comment (flip1995) [inline src/tools/clippy/tests/ui/into_iter_on_ref.fixed:14]: This test shouldn't be removed. If the lint now no longer triggers for it, the `ERROR` annotation should be removed and perhaps this line moved to the "should not lint" section of the test file, should that exist.

If it still triggers the lint, but shouldn't, than the Clippy lint code has to be fixed. If that should be the case, I can help you with that, if necessary.

**Review by flip1995 — CHANGES_REQUESTED** [review]
This Clippy test change is minor, but I want some clarification, before this PR gets merged.

File: src/tools/clippy/tests/ui/into_iter_on_ref.fixed
Comment (compiler-errors) [reply src/tools/clippy/tests/ui/into_iter_on_ref.fixed:14]: The lint no longer applies, since it no longer is redundant. 

In its stead, we now issue an edition migration warning BOXED_SLICE_INTO_ITER. I would need to remove the run-rustfix from this file, too, since the warning from this lint produces two suggestions which are mutually incompatible with each other.

Comment (flip1995) [reply src/tools/clippy/tests/ui/into_iter_on_ref.fixed:14]: So the Clippy lint no longer triggers? That's good.

Thanks for the explanation. I should've read more of the PR changes to find out myself. In that case removing this test case is the best thing to do.

**Review by flip1995 — APPROVED** [review]
No blockers from the Clippy side. Thanks for explaining!

File: compiler/rustc_lint/src/shadowed_into_iter.rs
Comment (WaffleLapkin) [inline compiler/rustc_lint/src/shadowed_into_iter.rs:83]:
```suggestion
        if cx.typeck_results().type_dependent_def_id(expr.hir_id) != cx.tcx.lang_items().into_iter_fn() {
            return;
        }
```

There is no need to re-wrap it in `Option` since `method_def_id` is not used anywhere else.

File: compiler/rustc_lint/src/shadowed_into_iter.rs
Comment (WaffleLapkin) [inline compiler/rustc_lint/src/shadowed_into_iter.rs:112]: instead of checking `idx`, you can do use `.next()` and then check if it's `is_array`, I think? This is a bit confusing and could probably use a comment, explaining what we are checking.

File: compiler/rustc_lint/src/shadowed_into_iter.rs
Comment (WaffleLapkin) [inline compiler/rustc_lint/src/shadowed_into_iter.rs:131]: Can you add a comment which describes what is checked here, exactly?

File: library/core/src/slice/iter.rs
Comment (WaffleLapkin) [inline library/core/src/slice/iter.rs:20]: I don't think this is controversial, but why is this needed?

File: tests/ui/iterators/into-iter-on-arrays-2018.stderr
Comment (WaffleLapkin) [inline tests/ui/iterators/into-iter-on-arrays-2018.stderr:61]: Why did this suggestion change?

File: tests/ui/iterators/into-iter-on-boxed-slices-2021.stderr
Comment (WaffleLapkin) [inline tests/ui/iterators/into-iter-on-boxed-slices-2021.stderr:56]: Shouldn't this suggest removing the call?

**Review by WaffleLapkin — CHANGES_REQUESTED** [review]
A few nit-picks and a question about suggestion that seems to be broken

File: compiler/rustc_lint/src/shadowed_into_iter.rs
Comment (compiler-errors) [reply compiler/rustc_lint/src/shadowed_into_iter.rs:83]: It protects against the `None == None` case, where we *do* want to return and not proceed into the code below.

File: compiler/rustc_lint/src/shadowed_into_iter.rs
Comment (compiler-errors) [reply compiler/rustc_lint/src/shadowed_into_iter.rs:112]: It seems a bit weird to do the check back over again though when we're already computing the right information here. I'll leave a comment clarifying what it's checking (and why), but I don't particularly want to change it.

File: compiler/rustc_lint/src/shadowed_into_iter.rs
Comment (compiler-errors) [reply compiler/rustc_lint/src/shadowed_into_iter.rs:131]: Yeah, I'll add a comment. It's just checking if the expr is `for _ in EXPR`. This was preexisting in the old version of the lint, for the record.

File: compiler/rustc_lint/src/shadowed_into_iter.rs
Comment (WaffleLapkin) [reply compiler/rustc_lint/src/shadowed_into_iter.rs:83]: Ah, right.

File: library/core/src/slice/iter.rs
Comment (compiler-errors) [reply library/core/src/slice/iter.rs:20]: Because otherwise we have an overlap with the `I: IntoIterator where I: Iterator` impl, via the `Box<I>: Iterator where I: Iterator` impl. I believe it has to do with the fact that `Box<I, A>` is fundamental, and it can't figure out that `Box<[I], A>` cannot be impl'd by a downstream crate. (I think it's the `A` parameter that it thinks can be substituted further with a downstream type...)

File: tests/ui/iterators/into-iter-on-arrays-2018.stderr
Comment (compiler-errors) [reply tests/ui/iterators/into-iter-on-arrays-2018.stderr:61]: :think: I think I'm doing the for loop desugaring computation wrong

File: compiler/rustc_lint/src/shadowed_into_iter.rs
Comment (WaffleLapkin) [reply compiler/rustc_lint/src/shadowed_into_iter.rs:112]: nvm, I misunderstood the check a bit. yeah, there is no need to recompute stuff, this is fine

File: library/core/src/slice/iter.rs
Comment (compiler-errors) [reply library/core/src/slice/iter.rs:20]: Yep, it's definitely the allocator param.

Comment (clarfonthey) [reply library/core/src/slice/iter.rs:20]: You're correct in the analysis here; this is what I had to add to make the library impl work

File: compiler/rustc_hir_typeck/src/method/prelude_edition_lints.rs
Comment (clarfonthey) [inline compiler/rustc_hir_typeck/src/method/prelude_edition_lints.rs:54]: Forgot to change the comment here

**Comment by Amanieu** [comment]
We discussed this in the libs-api meeting. We're happy with this design as it is, including re-using `vec::IntoIter`.

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated

We discussed this in the lang call on 2024-04-24.  People were in favor, and this is now in FCP.

**Comment by clarfonthey** [comment]
Thank you compiler-errors for taking this on while I had other stuff going on. Looking forward to this on stable. <3

File: tests/ui/iterators/into-iter-on-arrays-2018.stderr
Comment (compiler-errors) [reply tests/ui/iterators/into-iter-on-arrays-2018.stderr]: I fixed this

File: compiler/rustc_hir_typeck/src/method/prelude_edition_lints.rs
Comment (WaffleLapkin) [inline compiler/rustc_hir_typeck/src/method/prelude_edition_lints.rs:32]: Method is called `lint_dot_call_from_2018`, I don't think `2018` is still accurate?

File: compiler/rustc_hir_typeck/src/method/prelude_edition_lints.rs
Comment (WaffleLapkin) [reply compiler/rustc_hir_typeck/src/method/prelude_edition_lints.rs:54]: The comment is still the same?

File: library/alloc/src/boxed.rs
Comment (WaffleLapkin) [inline library/alloc/src/boxed.rs:2134]: Those are user facing docs, maybe we should write something that can be understood by normal users? (this can be a follow up)

**Review by WaffleLapkin — APPROVED** [review]
r=me with the nits

File: library/alloc/src/boxed.rs
Comment (compiler-errors) [reply library/alloc/src/boxed.rs:2134]: fff I thought I fixed these. I'll copy them over from the other PR for `Box<[T; N]>` where I think did.

**Comment by traviscross** [comment]
@rustbot labels +relnotes

We discussed in the edition call today that this should probably go in the release notes for Rust 1.80.

**Pull request merged**
